docs/devel/writing-good-e2e-tests.md

```diff
@@ -0,0 +1,168 @@
+log lots of irrelevant stuff either.
+1. **Ability to run in non-dedicated test clusters:** To reduce
+end-to-end delay and improve resource utilization when running e2e
+tests, we try, where possible, to run large numbers of tests in
+parallel against the same test cluster.  This means that:
+    1. you should avoid making any assumption (implicit or explicit)
+that your test is the only thing running against the cluster.  For
+example, making the assumption that your test can run a pod on every
+node in a cluster is not a safe assumption, as some other tests,
+running at the same time as yours, might have saturated one or more
+nodes in the cluster.  Similarly, running a pod in the system
+namespace, and assuming that that will increase the count of pods in
+the system namespace by one is not safe, as some other test might be
+creating or deleting pods in the system namespace at the same time as
+your test.  If you do legitimately need to write a test like that,
+make sure to label it "SERIAL" so that it's easy to identify, and not
+run in parallel with any other tests.
+    1. You should avoid doing things to the cluster that make it
+difficult for other tests to reliably do what they're trying to do, at
+the same time.  For example, rebooting nodes, disconnecting network
+interfaces, or upgrading cluster software as part of your test is
+likely to violate the assumptions that other tests might have made
+about a reasonably stable cluster environment.  If you need to write
+such tests, please label them as "DISRUPTIVE" so that it's easy to
+identify them, and not run them in parallel with other tests.
+1. **Speed of execution:** We have hundreds of e2e tests, some of
+which we run in serial, one after the other, in some cases.  If each
+test takes just a few minutes to run, that very quickly ads up to
+many, many hours of total execution time.  We try to keep such total
+execution time down to a few tens of minutes at most.  Therefore, try
+(very hard) to keep the execution time of your individual tests below
+2 minutes, ideally shorter than that.  Concretely, adding
+inappropriately long 'sleep' statements or other gratuitous waits to
+tests is a killer.  If under normal circumstances your pod enters the
+running state within 10 seconds, and 99.9% of the time within 30
+seconds, it would be gratuitous to wait 5 minutes for this to happen.
+Rather just fail after 30 seconds, with a clear error message as to
+why your test failed ("e.g. Pod x failed to become ready after 30
+seconds, it usually takes 10 seconds").  If you do have a truly
+legitimate reason for waiting longer than that, or writing a test
+which takes longer than 2 minutes to run, comment very clearly in the
+code why this is necessary, and label the test as "SLOW", so that it's
+easy to identify and avoid in test runs that are required to complete
+timeously (for example those that are run against every code
```

Comment:
FWIW, tests that are run against every submission are run in parallel, so I'm not clear on how concerned we are about tests that take ~10 min to run as long as they run in parallel.

cc @spxtr